osu.Game/Database/DetachedBeatmapStore.cs

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Game.Beatmaps;
+using Realms;
+
+namespace osu.Game.Database
+{
+    public partial class DetachedBeatmapStore : Component
+    {
+        private readonly ManualResetEventSlim loaded = new ManualResetEventSlim();
+
+        private readonly BindableList<BeatmapSetInfo> detachedBeatmapSets = new BindableList<BeatmapSetInfo>();
+
+        private IDisposable? realmSubscription;
+
+        private readonly Queue<OperationArgs> pendingOperations = new Queue<OperationArgs>();
+
+        [Resolved]
+        private RealmAccess realm { get; set; } = null!;
+
+        public IBindableList<BeatmapSetInfo> GetDetachedBeatmaps(CancellationToken? cancellationToken)
+        {
+            loaded.Wait(cancellationToken ?? CancellationToken.None);
+            return detachedBeatmapSets.GetBoundCopy();
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            realmSubscription = realm.RegisterForNotifications(r => r.All<BeatmapSetInfo>().Where(s => !s.DeletePending && !s.Protected), beatmapSetsChanged);
+        }
+
+        private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeSet? changes)
+        {
+            if (changes == null)
+            {
+                if (detachedBeatmapSets.Count > 0 && sender.Count == 0)
+                {
+                    // Usually we'd reset stuff here, but doing so triggers a silly flow which ends up deadlocking realm.
+                    // Additionally, user should not be at song select when realm is blocking all operations in the first place.
+                    //
+                    // Note that due to the catch-up logic below, once operations are restored we will still be in a roughly
+                    // correct state. The only things that this return will change is the carousel will not empty *during* the blocking
+                    // operation.
+                    return;
+                }
+
+                // Detaching beatmaps takes some time, so let's make sure it doesn't run on the update thread.
+                var frozenSets = sender.Freeze();
+
+                Task.Factory.StartNew(() =>
+                {
+                    realm.Run(_ =>
+                    {
+                        var detached = frozenSets.Detach();
+
+                        detachedBeatmapSets.Clear();
+                        detachedBeatmapSets.AddRange(detached);
+                        loaded.Set();
+                    });
+                }, TaskCreationOptions.LongRunning);
```

Comment:
No error handling here whatsoever worries me that if if this ever fails (for freak accident reasons) it will deadlock the game because `loaded` will never get set. Would need to test.

Reply:
Added minimal error handling and deadlock avoidance.